### PR TITLE
Editor.TokenColorCustomizations

### PR DESCRIPTION
**Pull request by hoovercj**
Addresses #27894

theme_overrides1

Caveat: colors update quickly, but font style does not. It will only update font style if I modify an affected token as shown in the gif below. Any ideas on what might cause that behavior?

theme_overrides2

Also, this pack does not include defining the config schema

### CODE REVIEW

**Comment by hoovercj** [comment]
@aeschli

File: src/vs/workbench/services/themes/electron-browser/workbenchThemeService.ts [hunk 1]
```diff
@@ -514,21 +527,61 @@ export class WorkbenchThemeService implements IWorkbenchThemeService {
...
-		let newColorCustomizations = this.configurationService.lookup<IColorCustomizations>(CUSTOM_COLORS_SETTING).value || {};
+		let newColorCustomizations = this.configurationService.lookup<IColorCustomizations>(CUSTOM_WORKBENCH_COLORS_SETTING).value || {};
 		let newColorIds = Object.keys(newColorCustomizations);
 		if (newColorIds.length === 0) {
 			newColorCustomizations = this.configurationService.lookup<IColorCustomizations>(DEPRECATED_CUSTOM_COLORS_SETTING).value || {};
 			newColorIds = Object.keys(newColorCustomizations);
 		}
-		if (this.hasCustomizationChanged(newColorCustomizations, newColorIds)) {
+
+		let newRules = this.getTokenColorCustomizations();
```
Comment (aeschli) [inline src/vs/workbench/services/themes/electron-browser/workbenchThemeService.ts]: Setting change notifications can happen quite often, cause by any change to settings. Mostly the change does not affect the custom colors. For that I tried to optimize the test for changes and avoid the creation of new data structures unless there is a change to the colors.
I'd suggest to not construct the rules array as in getTokenColorCustomizations, but just compare the old to the new settings.
The construction of the rules (as done in getTokenColorCustomizations) could be done inside setCustomTokenColors, like it is done with custom colors.

**Comment by aeschli** [comment]
Thanks @hoovercj , very nice, it's exactly what I have imagined!
I added some comments. It's a bit tight for the June release as it is the last day of development for that release. I'd suggest to commit it for July so we can get insiders feedback.

About the issue of fontstyle changes not triggering a refresh: This sounds like a bug in the editor.

**Comment by hoovercj** [comment]
I absolutely agree with waiting for the next release, @aeschli. We've waited this long, what's one more month? :-)

You said that you added comments but I can't see any.

File: src/vs/workbench/services/themes/electron-browser/themeCompatibility.ts [hunk 2]
```diff
@@ -26,12 +26,18 @@ function addSettingMapping(settingId: string, colorId: string) {
...
 			if (!settings) {
-				rule.settings = {};
+				rule.settings = {}; // make settings an empty object if it doesn't exist
 			} else {
+				// * For each key in settings (foreground, background, fontstyle)
+				// * If the key is in a settingToColorIdMapping, get the color from the setting
+				//   then find all the colorIds mapped to that setting, and set those to the color
+				//   It is not yet clear to me what this looks like in practice
+				// * Remove all keys that aren't foreground or background
+				//   Why remove fontstyle?
```
Comment (hoovercj) [inline src/vs/workbench/services/themes/electron-browser/themeCompatibility.ts]: When I go back in to remove my "question" comments, should I then also change `if (key !== 'foreground' && key !== 'background')` to `if (key !== 'foreground' && key !== 'background' && key !== 'fontStyle')`?

File: src/vs/workbench/services/themes/electron-browser/themeCompatibility.ts [hunk 3]
```diff
@@ -26,12 +26,18 @@ function addSettingMapping(settingId: string, colorId: string) {
 
 export function convertSettings(oldSettings: ITokenColorizationRule[], resultRules: ITokenColorizationRule[], resultColors: IColorMap): void {
 	for (let rule of oldSettings) {
-		resultRules.push(rule);
-		if (!rule.scope) {
+		resultRules.push(rule); // push the rule
+		if (!rule.scope) { // if it doesn't have a scope...?? why?
```
Comment (hoovercj) [inline src/vs/workbench/services/themes/electron-browser/themeCompatibility.ts]: Ok, that makes sense, thanks!

File: src/vs/workbench/services/themes/electron-browser/themeCompatibility.ts (hunk 2 above)
Comment (aeschli) [inline src/vs/workbench/services/themes/electron-browser/themeCompatibility.ts]: Yes

**Comment by hoovercj** [comment]
@aeschli is there any more action needed before merging this and getting insiders feedback?

**Comment by aeschli** [comment]
All looks good, thanks @hoovercj !

**Comment by khaosdoctor** [comment]
<div style="width:100%;height:0;padding-bottom:56%;position:relative;"></img></div>

**Comment by vytautas-pranskunas-** [comment]
Will it be whole scopes list?

**Comment by mscottfowler** [comment]
Hi, just downloaded the new VS Code July 2017. Do you have a list of the syntax tokens available (comments, strings) for the editor.tokenColorCustomizations user setting?

**Comment by hoovercj** [comment]
@mscottfowler the intellisense for the setting should show you all the options. For example, this is the gif from the release notes:

img

You can see that `comments`, `functions`, `keywords`, `numbers`, `strings`, `types`, and `variables` are options. If you use any of these, you'll change all comments, functions, etc. to the color you set.

If you want to be more specific, you can set ANY scope you want in the `textMateRules` section. This, too, is shown in the release notes. 

img2

To find the scopes that you need to use, open the command pallette and search "TM Scopes"
image

Then click anywhere in a file that you want to see what scopes are valid there.

image

I hope that helps!

**Comment by mscottfowler** [comment]
Thanks, Cody. Much appreciated!


Scott Fowler
Designer/Front-end Developer
505-985-1960
http://www.madjaybird.com/

**Comment by Mac-DevOSYandex** [comment]
And WHY O Why.. has Microsoft VS Code not implemented this as a feature that's part of the customization?!

**Comment by hoovercj** [comment]
@Mac-DevOSYandex This is a merged pull request so could you elaborate on what you mean? What do you believe is not implemented?

**Comment by Mac-DevOSYandex** [comment]
Hey Cody Hoover,

Many thanks for getting back to me.

Here's the pun, $200 for cursive font and yet with few code changes to vs code main css we can achieve cursive font in code.  I code in typescript most of the time and I code off line for a great deal of the time, therefore, I use vs code.

I came across this article (url below) and yes, it worked fine except, for the [unsportted] error message.
https://medium.com/@JeffHykin/how-to-get-cursive-fonts-in-vs-code-2018-without-operator-mono-220544fed93a

Once again many thanks..

Gio Gunn.

Sent with [ProtonMail](https://protonmail.com) Secure Email.

‐‐‐‐‐‐‐ Original Message ‐‐‐‐‐‐‐

**Pull request merged**